hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdb/util/ManagedSstFileReader.java

```diff
@@ -49,16 +49,37 @@ public class ManagedSstFileReader {
 
   private final Collection<String> sstFiles;
 
+  private volatile long estimatedTotalKeys = -1;
+
   public ManagedSstFileReader(final Collection<String> sstFiles) {
     this.sstFiles = sstFiles;
   }
 
   public static <T> Stream<T> getStreamFromIterator(ClosableIterator<T> itr) {
     final Spliterator<T> spliterator =
         Spliterators.spliteratorUnknownSize(itr, 0);
-    return StreamSupport.stream(spliterator, false).onClose(() -> {
-      itr.close();
-    });
+    return StreamSupport.stream(spliterator, false).onClose(itr::close);
+  }
+
+  public long getEstimatedTotalKeys() throws RocksDBException {
+    if (estimatedTotalKeys != -1) {
+      return estimatedTotalKeys;
+    }
+
+    long estimatedSize = 0;
+    synchronized (this) {
+      try (ManagedOptions options = new ManagedOptions()) {
+        for (String sstFile : sstFiles) {
+          SstFileReader fileReader = new SstFileReader(options);
+          fileReader.open(sstFile);
+          estimatedSize += fileReader.getTableProperties().getNumEntries();
+          estimatedSize += fileReader.getTableProperties().getNumDeletions();
+        }
+      }
+    }
+
+    estimatedTotalKeys = estimatedSize;
```

Comment:
Shouldn't we have this inside the synchronized block? And the synchronized block should also having the same if (estimatedTotalKeys != -1) {
      return estimatedTotalKeys;
    } to avoid redundant operations.

Reply:
1. It makes sense to move `estimatedTotalKeys = estimatedSize;` this in synchronized block. Will update it.
2. Regarding having double check `if (estimatedTotalKeys != -1)` inside the synchronized block, I agree but [findbug failed](https://github.com/apache/ozone/actions/runs/4813599053/jobs/8571917550) once when I did something similar to create singleton object. Another reason is that it doesn't matter that much because result is used by same thread. I'll add the check and see if findbug works. If does I'll update in next revision otherwise leave as it is.